File: grimdb/sdk/csharp/src/Models/PasswordEntry.cs
namespace Grimlocker.Models;

public sealed record PasswordEntry
{
    public string Title { get; init; } = "";
    public string Username { get; init; } = "";
    public string Password { get; init; } = "";
    public string Url { get; init; } = "";
    public string Notes { get; init; } = "";
    public string Id { get; init; } = "";

    public Dictionary<string, string> ToFields() => new()
    {
        ["username"] = Username, ["password"] = Password,
        ["url"] = Url, ["notes"] = Notes
    };

    public static PasswordEntry FromEntry(VaultEntry e) => new()
    {
        Id = e.Id, Title = e.Title,
        Username = e.Fields?.GetValueOrDefault("username") ?? "",
        Password = e.Fields?.GetValueOrDefault("password") ?? "",
        Url = e.Fields?.GetValueOrDefault("url") ?? "",
        Notes = e.Fields?.GetValueOrDefault("notes") ?? ""
    };
}

File: grimdb/sdk/swift/Sources/GrimlockerSDK/Models/PasswordEntry.swift
import Foundation

public struct PasswordEntry: Codable, Identifiable, Sendable {
    public var id: String
    public var title: String
    public var username: String
    public var password: String
    public var url: String
    public var notes: String

    enum CodingKeys: String, CodingKey {
        case id, title, username, password, url, notes
    }

    public init(id: String = "", title: String, username: String = "",
                password: String = "", url: String = "", notes: String = "") {
        self.id = id
        self.title = title
        self.username = username
        self.password = password
        self.url = url
        self.notes = notes
    }

    public var fields: [String: String] {
        [
            "username": username,
            "password": password,
            "url": url,
            "notes": notes,
        ]
    }

    public static func from(_ entry: VaultEntry) -> Self {
        .init(
            id: entry.id,
            title: entry.title,
            username: entry.field("username"),
            password: entry.field("password"),
            url: entry.field("url"),
            notes: entry.field("notes")
        )
    }
}
